guide/src/types.md

```diff
@@ -0,0 +1,101 @@
+# Overview of PyO3 Python object types
+
+On first glance, PyO3 provides a huge number of different types that can be used
+to wrap or refer to Python objects.  This page tries to give an overview of
+their meaning, with examples when each type is best used.
+
+* explain: pyclass
+* explain: GIL lifetime
+* explain: mutability?
+
+## Smart pointers
+
+### `PyObject`
+
+**Represents:** a GIL-independent reference to a Python object of unspecified
+type.
+
+**Used:** Whenever you want to carry around references to "some" Python object,
+without caring about a GIL lifetime.  For example, storing Python object
+references in a Rust struct that outlives the Python-Rust FFI boundary.
+
+Since `PyObject` does not carry a GIL lifetime, using most of its methods
+calling Python API will require a `Python` token.
+
+Can be cloned using Python reference counts with `.clone_ref()`.
+
+**Conversions:**
+
+- To `&PyAny`: `.as_ref(py)`
+- To `Py<ConcreteType>`: ???
```

Comment:
We can use `obj.as_ref(py).extract::<Py<ConcreteType>>`

Reply:
OK, that's good to know!